server/actions_specs.go

```diff
@@ -198,6 +203,46 @@ func (s *RegistryServer) getApiSpecRevision(ctx context.Context, name names.Spec
 	return message, nil
 }
 
+// GUnzippedBytes uncompresses a slice of bytes.
+func GUnzippedBytes(input []byte) ([]byte, error) {
+	buf := bytes.NewBuffer(input)
+	zr, err := gzip.NewReader(buf)
+	if err != nil {
+		return nil, err
+	}
+	return ioutil.ReadAll(zr)
+}
+
+// GetApiSpecContents handles the corresponding API request.
+func (s *RegistryServer) GetApiSpecContents(ctx context.Context, req *rpc.GetApiSpecContentsRequest) (*httpbody.HttpBody, error) {
+	var spec *rpc.ApiSpec
+	var err error
+	if name, err := names.ParseSpec(req.GetName()); err == nil {
+		spec, err = s.getApiSpec(ctx, name, rpc.View_FULL)
+	} else if name, err := names.ParseSpecRevision(req.GetName()); err == nil {
+		spec, err = s.getApiSpecRevision(ctx, name, rpc.View_FULL)
+	}
+	if err != nil {
+		return nil, err
+	}
+	if spec != nil {
+		contents := spec.Contents
+		contentType := spec.MimeType
+		if (strings.Contains(contentType, "+gzip")) {
+			contents, err = GUnzippedBytes(contents)
+			contentType = strings.Replace(contentType, "+gzip", "", 1)
+			if err != nil {
+				return nil, err
+			}
+		}
+		return &httpbody.HttpBody{
+			ContentType: contentType,
+			Data: contents,
+		}, nil
+	}
+	return nil, invalidArgumentError(fmt.Errorf("invalid resource name %q, must be an API spec or revision", req.GetName()))
```

Comment:
Is this error message correct? This only happens if getApiSpec or getApiSpecRevision return a nil spec and nil error, which seems to me more like an unknown internal error occurred.

Reply:
This is rearranged in the latest version.